{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "from pysam import VariantFile as vcf\n",
    "import operator\n",
    "from math import log2\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as dataframe\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import scipy\n",
    "import  os\n",
    "import os.path\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import csv\n",
    "import requests\n",
    "import re\n",
    "import json\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "split_race_region={'ACB': {},\n",
    " 'ASW': {},\n",
    " 'BEB': {'split_chr3/xav': [(98046342, 98273693)]},\n",
    " 'CDX': {'split_chr13/xai': [(55102733, 55354096)]},\n",
    " 'CEU': {'split_chr17/xaj': [(45670177, 46418024)]},\n",
    " 'CHB': {},\n",
    " 'CHS': {'split_chr3/xav': [(98046342, 98245327)]},\n",
    " 'CLM': {'split_chr3/xav': [(98046342, 98273693)]},\n",
    " 'ESN': {'split_chr5/xai': [(34311870, 34562837)]},\n",
    " 'FIN': {},\n",
    " 'GBR': {'split_chr17/xaj': [(45670177, 46418024)]},\n",
    " 'GIH': {'split_chr3/xav': [(98046342, 98276757)]},\n",
    " 'GWD': {},\n",
    " 'IBS': {'split_chr17/xaj': [(45836662, 46418024)]},\n",
    " 'ITU': {'split_chr17/xak': [(52575391, 52889013)]},\n",
    " 'JPT': {},\n",
    " 'KHV': {},\n",
    " 'LWK': {},\n",
    " 'MSL': {},\n",
    " 'MXL': {'split_chr3/xav': [(98046342, 98217489)]},\n",
    " 'PEL': {'split_chr11/xaf': [(23168664, 23461651)],\n",
    "  'split_chr3/xav': [(98046342, 98217489)]},\n",
    " 'PJL': {},\n",
    " 'PUR': {'split_chr3/xav': [(98046342, 98273693)]},\n",
    " 'STU': {'split_chr13/xai': [(55102733, 55443491)],\n",
    "  'split_chr3/xav': [(98046342, 98273693)]},\n",
    " 'TSI': {'split_chr17/xaj': [(45836662, 46418024)]},\n",
    " 'YRI': {}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene_info(chromosome, start, end):\n",
    "    query_xml = f\"\"\"\n",
    "    <Query virtualSchemaName=\"default\" formatter=\"TSV\" header=\"0\" uniqueRows=\"1\" count=\"\" datasetConfigVersion=\"0.6\">\n",
    "        <Dataset name=\"hsapiens_gene_ensembl\" interface=\"default\">\n",
    "            <Filter name=\"chromosome_name\" value=\"{chromosome}\"/>\n",
    "            <Filter name=\"start\" value=\"{start}\"/>\n",
    "            <Filter name=\"end\" value=\"{end}\"/>\n",
    "            <Attribute name=\"ensembl_gene_id\"/>\n",
    "            <Attribute name=\"external_gene_name\"/>\n",
    "            <Attribute name=\"description\"/>\n",
    "            <Attribute name=\"start_position\"/>\n",
    "            <Attribute name=\"end_position\"/>            \n",
    "            <Attribute name=\"strand\"/>\n",
    "        </Dataset>\n",
    "    </Query>\n",
    "    \"\"\"\n",
    "\n",
    "    biomart_url = \"http://www.ensembl.org/biomart/martservice?query=\"\n",
    "    response = requests.get(biomart_url + query_xml.strip())\n",
    "\n",
    "    if response.status_code != 200:\n",
    "        raise Exception(f\"Error fetching data from BioMart: {response.text}\")\n",
    "\n",
    "    genes = [line.split(\"\\t\") for line in response.text.strip().split(\"\\n\")]\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_gene_info(gene_list_one_record,chrnum):\n",
    "    result=[]\n",
    "    result.append(gene_list_one_record[0])\n",
    "    result.append(gene_list_one_record[1])\n",
    "    s=gene_list_one_record[3]\n",
    "    e=gene_list_one_record[4]\n",
    "    \n",
    "    pos=\"chr\"+chrnum+\":\"+str(s)+\"-\"+str(e)\n",
    "    result.append(pos)\n",
    "    result.append(gene_list_one_record[5])\n",
    "    result.append(gene_list_one_record[2])\n",
    "    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def extract_numbers(input_string):\n",
    "    return re.findall(r'\\d+', input_string)\n",
    "\n",
    "def ifgeneinside_region(complementary_region,gene_region):\n",
    "    \n",
    "    s=int(gene_region[0])\n",
    "    e=int(gene_region[1])\n",
    "     \n",
    "    if (complementary_region[0]<=s) and complementary_region[1]>=e:\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "13\n",
      "17\n",
      "3\n",
      "3\n",
      "5\n",
      "17\n",
      "3\n",
      "17\n",
      "17\n",
      "3\n",
      "11\n",
      "3\n",
      "3\n",
      "13\n",
      "3\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "allrecords=[]\n",
    "\n",
    "#non_mhc\n",
    "for race,split_ in split_race_region.items():\n",
    "    if split_ !={}:\n",
    "        if race!=\"All\":\n",
    "            for chr,arr in split_.items():\n",
    "                chr=chr.split(\"/\")[0].split(\"_\")[1]\n",
    "                chr=extract_numbers(chr)[0]\n",
    "                print(chr)\n",
    "                for a in arr:\n",
    "                    gene_list=fetch_gene_info(chromosome=chr,start=a[0]-1,end=a[1]-1)\n",
    "                    for gene_list_record in gene_list:\n",
    "                        if ifgeneinside_region(complementary_region=a,gene_region=[gene_list_record[3],gene_list_record[4]]):\n",
    "                            onerecord=[race,0]\n",
    "                            onerecord+=[\"chr\"+chr+\":\"+str(a[0])+\"-\"+str(a[1])]\n",
    "                            \n",
    "                            onerecord+=generate_gene_info(gene_list_one_record=gene_list_record,chrnum=chr)\n",
    "                            allrecords.append(onerecord)\n",
    "                        else:\n",
    "                            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generatedf(columns,allrecords):\n",
    "    dictforDF=dict()\n",
    "    for i in range(len(columns)):\n",
    "        midarr=[]\n",
    "        for record in allrecords:           \n",
    "            midarr.append(record[i])\n",
    "        dictforDF[columns[i]]=midarr\n",
    "    #dataframe(dictforDF).to_csv(\"csv2_05_09.csv\")\n",
    "    return dataframe(dictforDF)\n",
    "\n",
    "dfnew=generatedf(columns=[\"Race\",\"isMHC\",\"Position of disassortative mating region\",\"Gene ID\",\"Gene name\",\"Position of gene\",\"is_complement\",\"Gene description\"],allrecords=allrecords)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Race</th>\n",
       "      <th>isMHC</th>\n",
       "      <th>Position of disassortative mating region</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Position of gene</th>\n",
       "      <th>is_complement</th>\n",
       "      <th>Gene description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3:98046342-98273693</td>\n",
       "      <td>ENSG00000250750</td>\n",
       "      <td>OR5BM1P</td>\n",
       "      <td>chr3:98053374-98053739</td>\n",
       "      <td>1</td>\n",
       "      <td>olfactory receptor family 5 subfamily BM membe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3:98046342-98273693</td>\n",
       "      <td>ENSG00000213439</td>\n",
       "      <td>OR5AC1</td>\n",
       "      <td>chr3:98064472-98065396</td>\n",
       "      <td>1</td>\n",
       "      <td>olfactory receptor family 5 subfamily AC membe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3:98046342-98273693</td>\n",
       "      <td>ENSG00000196578</td>\n",
       "      <td>OR5AC2</td>\n",
       "      <td>chr3:98087173-98088102</td>\n",
       "      <td>1</td>\n",
       "      <td>olfactory receptor family 5 subfamily AC membe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3:98046342-98273693</td>\n",
       "      <td>ENSG00000249225</td>\n",
       "      <td></td>\n",
       "      <td>chr3:98099908-98148134</td>\n",
       "      <td>-1</td>\n",
       "      <td>novel transcript, antisense to OR5H1, OR5H14, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3:98046342-98273693</td>\n",
       "      <td>ENSG00000290819</td>\n",
       "      <td></td>\n",
       "      <td>chr3:98100793-98105672</td>\n",
       "      <td>1</td>\n",
       "      <td>novel transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>chr17:45836662-46418024</td>\n",
       "      <td>ENSG00000261575</td>\n",
       "      <td></td>\n",
       "      <td>chr17:46267412-46268694</td>\n",
       "      <td>1</td>\n",
       "      <td>C17orf58 pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>chr17:45836662-46418024</td>\n",
       "      <td>ENSG00000228696</td>\n",
       "      <td>ARL17B</td>\n",
       "      <td>chr17:46274784-46361766</td>\n",
       "      <td>-1</td>\n",
       "      <td>ADP ribosylation factor like GTPase 17B [Sourc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>chr17:45836662-46418024</td>\n",
       "      <td>ENSG00000289729</td>\n",
       "      <td></td>\n",
       "      <td>chr17:46283310-46286966</td>\n",
       "      <td>-1</td>\n",
       "      <td>bromodomain PHD finger transcription factor (B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>chr17:45836662-46418024</td>\n",
       "      <td>ENSG00000176681</td>\n",
       "      <td>LRRC37A</td>\n",
       "      <td>chr17:46292733-46338077</td>\n",
       "      <td>1</td>\n",
       "      <td>leucine rich repeat containing 37A [Source:HGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>chr17:45836662-46418024</td>\n",
       "      <td>ENSG00000265411</td>\n",
       "      <td>RN7SL656P</td>\n",
       "      <td>chr17:46319679-46319959</td>\n",
       "      <td>1</td>\n",
       "      <td>RNA, 7SL, cytoplasmic 656, pseudogene [Source:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Race  isMHC Position of disassortative mating region          Gene ID  \\\n",
       "0    BEB      0                   chr3:98046342-98273693  ENSG00000250750   \n",
       "1    BEB      0                   chr3:98046342-98273693  ENSG00000213439   \n",
       "2    BEB      0                   chr3:98046342-98273693  ENSG00000196578   \n",
       "3    BEB      0                   chr3:98046342-98273693  ENSG00000249225   \n",
       "4    BEB      0                   chr3:98046342-98273693  ENSG00000290819   \n",
       "..   ...    ...                                      ...              ...   \n",
       "205  TSI      0                  chr17:45836662-46418024  ENSG00000261575   \n",
       "206  TSI      0                  chr17:45836662-46418024  ENSG00000228696   \n",
       "207  TSI      0                  chr17:45836662-46418024  ENSG00000289729   \n",
       "208  TSI      0                  chr17:45836662-46418024  ENSG00000176681   \n",
       "209  TSI      0                  chr17:45836662-46418024  ENSG00000265411   \n",
       "\n",
       "     Gene name         Position of gene is_complement  \\\n",
       "0      OR5BM1P   chr3:98053374-98053739             1   \n",
       "1       OR5AC1   chr3:98064472-98065396             1   \n",
       "2       OR5AC2   chr3:98087173-98088102             1   \n",
       "3                chr3:98099908-98148134            -1   \n",
       "4                chr3:98100793-98105672             1   \n",
       "..         ...                      ...           ...   \n",
       "205             chr17:46267412-46268694             1   \n",
       "206     ARL17B  chr17:46274784-46361766            -1   \n",
       "207             chr17:46283310-46286966            -1   \n",
       "208    LRRC37A  chr17:46292733-46338077             1   \n",
       "209  RN7SL656P  chr17:46319679-46319959             1   \n",
       "\n",
       "                                      Gene description  \n",
       "0    olfactory receptor family 5 subfamily BM membe...  \n",
       "1    olfactory receptor family 5 subfamily AC membe...  \n",
       "2    olfactory receptor family 5 subfamily AC membe...  \n",
       "3    novel transcript, antisense to OR5H1, OR5H14, ...  \n",
       "4                                     novel transcript  \n",
       "..                                                 ...  \n",
       "205                                C17orf58 pseudogene  \n",
       "206  ADP ribosylation factor like GTPase 17B [Sourc...  \n",
       "207  bromodomain PHD finger transcription factor (B...  \n",
       "208  leucine rich repeat containing 37A [Source:HGN...  \n",
       "209  RNA, 7SL, cytoplasmic 656, pseudogene [Source:...  \n",
       "\n",
       "[210 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfnew"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfnew.to_csv(\"split.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mhc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
