{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pysam\n",
    "from pysam import VariantFile as vcf\n",
    "import operator\n",
    "from math import log2\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as dataframe\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import scipy\n",
    "import  os\n",
    "import os.path\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import csv\n",
    "from xml.etree import ElementTree as ET\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "\n",
    "all_mean_arr=-0.019077723390066942\n",
    "all_std_arr=0.006128439002438663\n",
    "mean_dict_different_race={'ACB': -0.014011151662888553,\n",
    " 'ASW': -0.012693412503283,\n",
    " 'BEB': -0.0038608773522230016,\n",
    " 'CDX': -0.000870975201046489,\n",
    " 'CEU': -0.002626550761399808,\n",
    " 'CHB': -0.000840617782511786,\n",
    " 'CHS': -0.0016030210248706835,\n",
    " 'CLM': -0.006497201413442441,\n",
    " 'ESN': -0.012407888605124987,\n",
    " 'FIN': -0.0013888123211964092,\n",
    " 'GBR': -0.0016421960521901954,\n",
    " 'GIH': -0.0029655045615120147,\n",
    " 'GWD': -0.013383054011719266,\n",
    " 'IBS': -0.0036274568161741316,\n",
    " 'ITU': -0.003697573961398361,\n",
    " 'JPT': -0.0004422965646677968,\n",
    " 'KHV': -0.0014824412579579188,\n",
    " 'LWK': -0.013214685221562708,\n",
    " 'MSL': -0.012902619299438558,\n",
    " 'MXL': -0.004064132377929528,\n",
    " 'PEL': -0.0032051883341473507,\n",
    " 'PJL': -0.004661546996727815,\n",
    " 'PUR': -0.008252100717713692,\n",
    " 'STU': -0.003961436794723387,\n",
    " 'TSI': -0.002358177545570631,\n",
    " 'YRI': -0.012978184004079407}\n",
    "std_dict_different_race={'ACB': 0.007679014864719199,\n",
    " 'ASW': 0.008041527656473924,\n",
    " 'BEB': 0.008061181937627181,\n",
    " 'CDX': 0.008574772155886829,\n",
    " 'CEU': 0.007805712926154353,\n",
    " 'CHB': 0.00858396197930932,\n",
    " 'CHS': 0.008210661522162982,\n",
    " 'CLM': 0.007855389049385354,\n",
    " 'ESN': 0.007528169549036397,\n",
    " 'FIN': 0.008233373830804186,\n",
    " 'GBR': 0.008280482586380011,\n",
    " 'GIH': 0.008318428884896655,\n",
    " 'GWD': 0.00741808675422558,\n",
    " 'IBS': 0.007946168606506706,\n",
    " 'ITU': 0.007944085877932789,\n",
    " 'JPT': 0.008520600208015637,\n",
    " 'KHV': 0.008457471872578115,\n",
    " 'LWK': 0.00775427047797345,\n",
    " 'MSL': 0.00788405338804666,\n",
    " 'MXL': 0.008244211463289736,\n",
    " 'PEL': 0.008220014590454098,\n",
    " 'PJL': 0.007728364636381623,\n",
    " 'PUR': 0.007789028974214059,\n",
    " 'STU': 0.007989081367465312,\n",
    " 'TSI': 0.008171487042712589,\n",
    " 'YRI': 0.007574843363674686}\n",
    "\n",
    "df_csv2=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/python/csv3_again/final_buchong_0722.csv\").values.tolist()\n",
    "\n",
    "def getmaxindices(arr,num):\n",
    "    max_values_and_indices = sorted(enumerate(arr), key=lambda x: x[1],reverse=True)[:num]\n",
    "    max_indices = [index for index, value in max_values_and_indices]    \n",
    "    return max_indices\n",
    "\n",
    "\n",
    "allrecords=[]\n",
    "#in csv2\n",
    "for record in df_csv2:\n",
    "    if type(record[6])!=float:\n",
    "\n",
    "        if record[1]==\"All\":\n",
    "            if record[2]==0:#no mhc csvfiles                        \n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/csvfiles/\")\n",
    "                for filename in filenames:\n",
    "                    if filename.startswith(\"pos_prob_split_\"+chr+\"_\"):\n",
    "                        df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/csvfiles/\"+filename)  \n",
    "\n",
    "                        allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                        allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                        if allrigion_P_top_1000:\n",
    "                            indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                            p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                            pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "    \n",
    "                            \n",
    "                            for j in range(len(p_top_100)):\n",
    "                                pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                                onerecord=[]\n",
    "                                onerecord+=record[1:8]\n",
    "                                onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                                allrecords.append(onerecord)\n",
    "                                \n",
    "\n",
    "                        else:\n",
    "                            continue\n",
    "\n",
    "            else:\n",
    "                position_of_gene=record[6]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/csvfiles/posmhc.csv\")\n",
    "                allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                if allrigion_P_top_1000:\n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "\n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "               \n",
    "\n",
    "        else:#race!=all\n",
    "            if record[2]==0:#no mhc\n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/race/\")\n",
    "                for filename in filenames:\n",
    "                    if filename.startswith(\"split_\"+chr+\"_\"):\n",
    "                        df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/race/\"+filename)  \n",
    "                        allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                        allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                        if allrigion_P_top_1000:\n",
    "                            indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                            p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                            pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "                            for j in range(len(p_top_100)):\n",
    "                                pvalue=round((1-stats.norm.cdf(p_top_100[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                                onerecord=[]\n",
    "                                onerecord+=record[1:8]\n",
    "                                onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                                allrecords.append(onerecord)\n",
    "                        else:\n",
    "                            continue\n",
    "                    else:\n",
    "                        continue\n",
    "            else:#ismhc different race\n",
    "                position_of_gene=record[6]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/race/mhc.csv\")\n",
    "                allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                if allrigion_P_top_1000:  \n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]              \n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "   \n",
    "    else:\n",
    "\n",
    "        onerecord=[]\n",
    "        onerecord+=record[1:8]\n",
    "        onerecord+=[\"null\",\"null\",\"null\"]\n",
    "        allrecords.append(onerecord)\n",
    "\n",
    "\n",
    "\n",
    "#generatedf\n",
    "def generatedf(columns,allrecords):\n",
    "    dictforDF=dict()\n",
    "    for i in range(len(columns)):\n",
    "        midarr=[]\n",
    "        for record in allrecords:           \n",
    "            midarr.append(record[i])\n",
    "        dictforDF[columns[i]]=midarr\n",
    "    dataframe(dictforDF).to_csv(\"/data2/wangxuedong/mhc_test_data/python/csv3_again/csv3_buchong_withoutcorr.csv\")\n",
    "    return dataframe(dictforDF)\n",
    "\n",
    "\n",
    "dfnew=generatedf(columns=[\"Race\",\"isMHC\",\"Position of disassortative mating region\",\"Gene ID\",\"Gene name\",\"Position of gene\",\"is_complement\",\"Position of SNP\",\"P\",\"p-value\"],allrecords=allrecords)\n",
    "\n",
    "\n",
    "pvalue_list=dfnew[\"p-value\"].tolist()\n",
    "corr_pvals = multipletests(pvalue_list, method='fdr_bh')[1]\n",
    "dfnew = dfnew.assign(corr_pvals=corr_pvals)\n",
    "\n",
    "dfnew.to_csv(\"/data2/wangxuedong/mhc_test_data/python/csv3_again/csv3_buchong_withcorr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mhc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
