{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pysam\n",
    "from pysam import VariantFile as vcf\n",
    "import operator\n",
    "from math import log2\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as dataframe\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import scipy\n",
    "import  os\n",
    "import os.path\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import csv\n",
    "from xml.etree import ElementTree as ET\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "all_mean_arr=0.03411387460093794\n",
    "all_std_arr=0.01334116014092687\n",
    "mean_dict_different_race={'ACB': 0.046621686769679045,\n",
    " 'ASW': 0.0465892685112964,\n",
    " 'BEB': 0.035970363185117774,\n",
    " 'CDX': 0.03266259792789275,\n",
    " 'CEU': 0.03488434818325704,\n",
    " 'CHB': 0.03285897897059889,\n",
    " 'CHS': 0.032796236687928325,\n",
    " 'CLM': 0.036503986698316906,\n",
    " 'ESN': 0.04643492431530403,\n",
    " 'FIN': 0.034935842703769975,\n",
    " 'GBR': 0.03498406394197642,\n",
    " 'GIH': 0.035593525252319486,\n",
    " 'GWD': 0.045836194805001416,\n",
    " 'IBS': 0.035140910353782764,\n",
    " 'ITU': 0.0347164844670136,\n",
    " 'JPT': 0.03273063391657775,\n",
    " 'KHV': 0.033023382087129426,\n",
    " 'LWK': 0.0466350005070252,\n",
    " 'MSL': 0.047257776857587634,\n",
    " 'MXL': 0.03529323182476227,\n",
    " 'PEL': 0.03222809053641309,\n",
    " 'PJL': 0.03450825905794522,\n",
    " 'PUR': 0.037936602232609126,\n",
    " 'STU': 0.03454596363552346,\n",
    " 'TSI': 0.03508890911947475,\n",
    " 'YRI': 0.04649868546958651}\n",
    "std_dict_different_race={'ACB': 0.01660837078934563,\n",
    " 'ASW': 0.016852430054156438,\n",
    " 'BEB': 0.016339814570213686,\n",
    " 'CDX': 0.01657929168371163,\n",
    " 'CEU': 0.01615285680476383,\n",
    " 'CHB': 0.0167769938917591,\n",
    " 'CHS': 0.016361466363591634,\n",
    " 'CLM': 0.016076712212741314,\n",
    " 'ESN': 0.016439364809682746,\n",
    " 'FIN': 0.016395558138330903,\n",
    " 'GBR': 0.01654186443581865,\n",
    " 'GIH': 0.016487954511321803,\n",
    " 'GWD': 0.01637735033664222,\n",
    " 'IBS': 0.01625740426618882,\n",
    " 'ITU': 0.015999114837619923,\n",
    " 'JPT': 0.016777856667394144,\n",
    " 'KHV': 0.016805860157616894,\n",
    " 'LWK': 0.01656542896110204,\n",
    " 'MSL': 0.0168657687428544,\n",
    " 'MXL': 0.01631757749279766,\n",
    " 'PEL': 0.015903810806753623,\n",
    " 'PJL': 0.015785059411911464,\n",
    " 'PUR': 0.016016827217755162,\n",
    " 'STU': 0.01615897181252148,\n",
    " 'TSI': 0.016499539576161064,\n",
    " 'YRI': 0.01656520483641468}\n",
    "\n",
    "df_csv2=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/find_region/dis_csv/tables3_01part.csv\").values.tolist()\n",
    "\n",
    "def getmaxindices(arr,num):\n",
    "    max_values_and_indices = sorted(enumerate(arr), key=lambda x: x[1],reverse=True)[:num]\n",
    "    max_indices = [index for index, value in max_values_and_indices]    \n",
    "    return max_indices\n",
    "\n",
    "\n",
    "allrecords=[]\n",
    "#in csv2\n",
    "for record in df_csv2:\n",
    "    if type(record[6])!=float:\n",
    "\n",
    "        if record[1]==\"All\":\n",
    "            if record[2]==0:#no mhc csvfiles                        \n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                #filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/first_review/csvfiles_integration/\")\n",
    "                # for filename in filenames:\n",
    "                #     if filename.startswith(\"pos_prob_split_\"+chr+\"_\"):\n",
    "                df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/csvfiles_integration/\"+chr+\".csv\")  \n",
    "\n",
    "                allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                if allrigion_P_top_1000:\n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "\n",
    "                    \n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                        \n",
    "\n",
    "                else:\n",
    "                    continue\n",
    "\n",
    "            else:\n",
    "                position_of_gene=record[6]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/csvfiles/mhc.csv\")\n",
    "                allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                if allrigion_P_top_1000:\n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "\n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "               \n",
    "\n",
    "        else:#race!=all\n",
    "            if record[2]==0:#no mhc\n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                # filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/race/\")\n",
    "                # for filename in filenames:\n",
    "                #     if filename.startswith(\"split_\"+chr+\"_\"):\n",
    "                df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/race_integration/\"+chr+\".csv\")  \n",
    "                allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                if allrigion_P_top_1000:\n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "              \n",
    "            else:#ismhc different race\n",
    "                position_of_gene=record[6]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/race/posmhc.csv\")\n",
    "                allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                if allrigion_P_top_1000:  \n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]              \n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "   \n",
    "    else:\n",
    "\n",
    "        onerecord=[]\n",
    "        onerecord+=record[1:8]\n",
    "        onerecord+=[\"null\",\"null\",\"null\"]\n",
    "        allrecords.append(onerecord)\n",
    "\n",
    "\n",
    "\n",
    "#generatedf\n",
    "def generatedf(columns,allrecords):\n",
    "    dictforDF=dict()\n",
    "    for i in range(len(columns)):\n",
    "        midarr=[]\n",
    "        for record in allrecords:           \n",
    "            midarr.append(record[i])\n",
    "        dictforDF[columns[i]]=midarr\n",
    "    dataframe(dictforDF).to_csv(\"tables4_withoutcorr.csv\")\n",
    "    return dataframe(dictforDF)\n",
    "\n",
    "\n",
    "dfnew=generatedf(columns=[\"Race\",\"isMHC\",\"Position of disassortative mating region\",\"Gene ID\",\"Gene name\",\"Position of gene\",\"is_complement\",\"Position of SNP\",\"P\",\"p-value\"],allrecords=allrecords)\n",
    "\n",
    "\n",
    "pvalue_list=dfnew[\"p-value\"].tolist()\n",
    "corr_pvals = multipletests(pvalue_list, method='fdr_bh')[1]\n",
    "dfnew = dfnew.assign(corr_pvals=corr_pvals)\n",
    "\n",
    "dfnew.to_csv(\"tables4_withcorr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df_csv2=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/find_region/dis_csv/tables3_01part.csv\").values.tolist()\n",
    "for record in df_csv2:\n",
    "    if type(record[6])==float:\n",
    "        print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mhc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
