{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pysam\n",
    "from pysam import VariantFile as vcf\n",
    "import operator\n",
    "from math import log2\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as dataframe\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import scipy\n",
    "import  os\n",
    "import os.path\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy import stats\n",
    "import csv\n",
    "from xml.etree import ElementTree as ET\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "all_mean_arr=0.008131350120614151\n",
    "all_std_arr=0.003374720423931657\n",
    "mean_dict_different_race={'ACB': 0.043905219284227484,\n",
    " 'ASW': 0.061747415774036245,\n",
    " 'BEB': 0.042622549524204016,\n",
    " 'CDX': 0.05629086218276262,\n",
    " 'CEU': 0.03350899999290784,\n",
    " 'CHB': 0.05201986349516006,\n",
    " 'CHS': 0.03643899713354041,\n",
    " 'CLM': 0.04224577966848677,\n",
    " 'ESN': 0.03623741211034457,\n",
    " 'FIN': 0.053129961026517944,\n",
    " 'GBR': 0.05656052651694239,\n",
    " 'GIH': 0.05157971151605565,\n",
    " 'GWD': 0.03170015029229307,\n",
    " 'IBS': 0.0371240314050134,\n",
    " 'ITU': 0.050724822118831774,\n",
    " 'JPT': 0.0517456479295656,\n",
    " 'KHV': 0.04564560273080832,\n",
    " 'LWK': 0.04973674801295715,\n",
    " 'MSL': 0.04960771674064523,\n",
    " 'MXL': 0.05377595974774834,\n",
    " 'PEL': 0.045806236523102895,\n",
    " 'PJL': 0.040244478916681585,\n",
    " 'PUR': 0.04003893507415216,\n",
    " 'STU': 0.048497323483452925,\n",
    " 'TSI': 0.05007660537347943,\n",
    " 'YRI': 0.031315221824936575}\n",
    "std_dict_different_race={'ACB': 0.0035837115206611876,\n",
    " 'ASW': 0.0044435636072078216,\n",
    " 'BEB': 0.003260312673626174,\n",
    " 'CDX': 0.00391015689396564,\n",
    " 'CEU': 0.002766896630079425,\n",
    " 'CHB': 0.0037059827211937194,\n",
    " 'CHS': 0.002921456474220746,\n",
    " 'CLM': 0.003272057962121683,\n",
    " 'ESN': 0.0031803272910866064,\n",
    " 'FIN': 0.0038008600285356704,\n",
    " 'GBR': 0.003936807447491629,\n",
    " 'GIH': 0.0037221796184063703,\n",
    " 'GWD': 0.0029073393803639483,\n",
    " 'IBS': 0.0029858040112478645,\n",
    " 'ITU': 0.003597951675686167,\n",
    " 'JPT': 0.003708599258289701,\n",
    " 'KHV': 0.00343805410624286,\n",
    " 'LWK': 0.003848934277722462,\n",
    " 'MSL': 0.003934844393069718,\n",
    " 'MXL': 0.003865357254429113,\n",
    " 'PEL': 0.003392906748737866,\n",
    " 'PJL': 0.003126523821241254,\n",
    " 'PUR': 0.0031949284883587385,\n",
    " 'STU': 0.00353413807778532,\n",
    " 'TSI': 0.0036716230350532964,\n",
    " 'YRI': 0.002910364879924007}\n",
    "\n",
    "df_csv2=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/find_region/pos_csv/tables5_part0.csv\").values.tolist()\n",
    "\n",
    "def getmaxindices(arr,num):\n",
    "    max_values_and_indices = sorted(enumerate(arr), key=lambda x: x[1],reverse=True)[:num]\n",
    "    max_indices = [index for index, value in max_values_and_indices]    \n",
    "    return max_indices\n",
    "\n",
    "mhcrace_record=[]\n",
    "allrecords=[]\n",
    "#in csv2\n",
    "for record in df_csv2:\n",
    "    if type(record[6])!=float:\n",
    "        # print(\"record\",record)\n",
    "        # print(\"record6\",record[6])\n",
    "        if record[1]==\"All\":\n",
    "            if record[2]==0:#no mhc csvfiles   all                     \n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                if chr==\"chr6\":\n",
    "                    filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_csvfiles_inte/\")\n",
    "                    for filename in filenames:\n",
    "                        if filename.startswith(\"chr6_\"):\n",
    "                            df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_csvfiles_inte/\"+filename)  \n",
    "\n",
    "                            allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                            allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                            if allrigion_P_top_1000:\n",
    "                                indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                                p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                                pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "        \n",
    "                                \n",
    "                                for j in range(len(p_top_100)):\n",
    "                                    pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                                    onerecord=[]\n",
    "                                    onerecord+=record[1:8]\n",
    "                                    onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                                    allrecords.append(onerecord)\n",
    "                                    \n",
    "\n",
    "                            else:\n",
    "                                continue\n",
    "                        else:\n",
    "                            continue\n",
    "                else:\n",
    "                    \n",
    "                    df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_csvfiles_inte/\"+chr+\".csv\")  \n",
    "\n",
    "                    allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                    allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                    if allrigion_P_top_1000:\n",
    "                        indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                        p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                        pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "\n",
    "                        \n",
    "                        for j in range(len(p_top_100)):\n",
    "                            pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                            onerecord=[]\n",
    "                            onerecord+=record[1:8]\n",
    "                            onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                            allrecords.append(onerecord)\n",
    "                            \n",
    "\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "            else:#mhc all\n",
    "                position_of_gene=record[6]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_csvfiles/posmhc.csv\")\n",
    "                allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"P\"].to_list()\n",
    "                if allrigion_P_top_1000:\n",
    "                    # indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    # p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    # pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "                    # mean_100_P=np.mean(p_top_100)\n",
    "                    # std_100_P=np.std(p_top_100, ddof=1)\n",
    "                    # pvalue=round((1-stats.norm.cdf(mean_100_P, all_mean_arr, all_std_arr)),15)\n",
    "                    indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                    p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                    pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "                    for j in range(len(p_top_100)):\n",
    "                        pvalue=round((1-stats.norm.cdf(p_top_100[j], all_mean_arr, all_std_arr)),15)\n",
    "                        onerecord=[]\n",
    "                        onerecord+=record[1:8]\n",
    "                        onerecord+=[pos_top_100[j],pos_top_100[j],pvalue]\n",
    "                        allrecords.append(onerecord)\n",
    "                else:\n",
    "                    continue\n",
    "                # print(\"mhc region\",record)\n",
    "\n",
    "        else:#race!=all\n",
    "            if record[2]==0:#no mhc\n",
    "                position_of_gene=record[6]\n",
    "                chr=position_of_gene.split(\":\")[0]\n",
    "                start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "\n",
    "                end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                if chr==\"chr6\":\n",
    "                    filenames=os.listdir(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_race_integration/\")\n",
    "                    for filename in filenames:\n",
    "                        if filename.startswith(\"chr6_\"):\n",
    "                            df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_race_integration/\"+filename)  \n",
    "                            allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                            allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                            if allrigion_P_top_1000:\n",
    "                                indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                                p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                                pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "                                for j in range(len(p_top_100)):\n",
    "                                    pvalue=round((1-stats.norm.cdf(p_top_100[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                                    onerecord=[]\n",
    "                                    onerecord+=record[1:8]\n",
    "                                    onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                                    allrecords.append(onerecord)\n",
    "                            else:\n",
    "                                continue\n",
    "                        else:\n",
    "                            continue\n",
    "                else:\n",
    "                    df_split=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/similar_region_race_integration/\"+chr+\".csv\")  \n",
    "\n",
    "                    allpos=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                    allrigion_P_top_1000=df_split[(df_split[\"Unnamed: 0\"]>=start_position) & (df_split[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                    if allrigion_P_top_1000:\n",
    "                        indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                        p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                        pos_top_100=[allpos[i] for i in indices_top_100]\n",
    "\n",
    "                        \n",
    "                        for j in range(len(p_top_100)):\n",
    "                            pvalue=round((1-stats.norm.cdf(p_top_100[j], mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                            onerecord=[]\n",
    "                            onerecord+=record[1:8]\n",
    "                            onerecord+=[pos_top_100[j],p_top_100[j],pvalue]\n",
    "                            allrecords.append(onerecord)\n",
    "                            \n",
    "\n",
    "                    else:\n",
    "                        continue\n",
    "            else:#ismhc different race\n",
    "                # position_of_gene=record[6]\n",
    "                # start_position=int(position_of_gene.split(\":\")[1].split(\"-\")[0])\n",
    "                # end_position=int(position_of_gene.split(\":\")[1].split(\"-\")[1])\n",
    "                # df_mhc=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/race/mhc.csv\")\n",
    "                # allpos=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][\"Unnamed: 0\"].to_list()\n",
    "                # allrigion_P_top_1000=df_mhc[(df_mhc[\"Unnamed: 0\"]>=start_position) & (df_mhc[\"Unnamed: 0\"]<=end_position)][record[1]].to_list()\n",
    "                # if allrigion_P_top_1000:  \n",
    "                #     indices_top_100=getmaxindices(allrigion_P_top_1000,100)\n",
    "                #     p_top_100=[allrigion_P_top_1000[i] for i in indices_top_100]\n",
    "                #     pos_top_100=[allpos[i] for i in indices_top_100]              \n",
    "                #     for j in range(len(allrigion_P_top_1000)):\n",
    "                #         pvalue=round((1-stats.norm.cdf(allrigion_P_top_1000[j],mean_dict_different_race[record[1]],std_dict_different_race[record[1]])),15)\n",
    "                #         onerecord=[]\n",
    "                #         onerecord+=record[1:8]\n",
    "                #         onerecord+=[allpos[j],allrigion_P_top_1000[j],pvalue]\n",
    "                #         allrecords.append(onerecord)\n",
    "                # else:\n",
    "                #     continue\n",
    "                mhcrace_record.append(record)\n",
    "                print(\"mhc record in race\",record)\n",
    "    else:\n",
    "        # print(\"record\",record)\n",
    "        # print(\"record6\",record[6])\n",
    "        onerecord=[]\n",
    "        onerecord+=record[1:8]\n",
    "        onerecord+=[\"null\",\"null\",\"null\"]\n",
    "        allrecords.append(onerecord)\n",
    "\n",
    "\n",
    "\n",
    "#generatedf\n",
    "def generatedf(columns,allrecords):\n",
    "    dictforDF=dict()\n",
    "    for i in range(len(columns)):\n",
    "        midarr=[]\n",
    "        for record in allrecords:           \n",
    "            midarr.append(record[i])\n",
    "        dictforDF[columns[i]]=midarr\n",
    "    dataframe(dictforDF).to_csv(\"tables6_nocorr_xiu.csv\")\n",
    "    return dataframe(dictforDF)\n",
    "\n",
    "\n",
    "dfnew=generatedf(columns=[\"Race\",\"isMHC\",\"Position of disassortative mating region\",\"Gene ID\",\"Gene name\",\"Position of gene\",\"is_complement\",\"Position of SNP\",\"P\",\"p-value\"],allrecords=allrecords)\n",
    "\n",
    "\n",
    "pvalue_list=dfnew[\"p-value\"].tolist()\n",
    "corr_pvals = multipletests(pvalue_list, method='fdr_bh')[1]\n",
    "dfnew = dfnew.assign(corr_pvals=corr_pvals)\n",
    "\n",
    "dfnew.to_csv(\"tables6_withcorr_xiu.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_csv2=pd.read_csv(\"/data2/wangxuedong/mhc_test_data/first_review/find_region/pos_csv/tables5_part0.csv\").values.tolist()\n",
    "for record in df_csv2:\n",
    "    if type(record[6])==float:\n",
    "        print(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhcrace_record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mhc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
